### DISCUSSION TITLE
[Feature] Allow to call callbacks before responding

### DISCUSSION THREAD

**Discussion by sebastien-savalle** [mockoon/mockoon#1287] [Ideas]
Hello,

First, I'd like to thank you for all your hard work.
The introduction of callbacks is a great achievement but sometimes It doesn't fit our needs.

I found at least two use cases where I would like to call a callback before answering to the request:

1. It would be nice to be able to reuse data provided in the response callback in the response of the initial request.
2. We have some use case where we must call another endpoint of the originator of the request before replying.

Would it be possible to enhance callbacks with such options ?

Thanks a lot

**Comment by 255kb** [comment]
Hi,

> First, I'd like to thank you for all your hard work.
Thank you!

I think this is a request similar to #1286. 

My opinion is that this should be a different feature than the callbacks, more like some asynchronous data fetching prior to answering the request. Nothing like this is planned for now and I think it defeats a bit the purpose of a self-contained API mock if it depends on external data. 
I'm happy to re-think about it if this discussion gather some interest. 

Thanks

**Reply by sebastien-savalle** [reply]
Hi,

I agree that this is something a little bit different than callbacks. But it should be possible to implement it quickly by reusing the callback approach. At least the second point when we just want to call the callbacks before answering without reusing the response.

Thanks,

**Reply by 255kb** [reply]
Technically, callbacks are called before the response is sent, we are just not waiting for the answer. 

The issue is not really "is it easy to do?" but rather will it contribute to the app cluttering, make the interface more complex, make the code harder to maintain, etc. I would be happy to add all the requested features, but now that the user base tend to be quite big, I prefer to say no first and re-evaluate if needed.

**Reply by sebastien-savalle** [reply]
Not really, I have tried to delay the response with a latency set to a few seconds in order to find a workaround. In that case the callback are sent when the response latency expired.

**Reply by 255kb** [reply]
Yes, you are right. I didn't think about the route latency. The purpose of the callbacks were also to mimick the equivalent webhooks feature in other tools like Wiremock. Something that you call after something else is done.
I'm not sure we want to change this behavior.

**Reply by sgarouachi** [reply]
Maybe enhancing the callback feature by introducing a configuration of its desired behavior (as below) might help;
- **after the response (default)**: callback requests are sent after the response
- **before the response**: callback requests are sent before the response
